producer.go

```diff
@@ -116,11 +116,32 @@ func (p *Producer) Start() {
 	p.started = true
 }
 
-// Stop gracefully shutsdown the producer, cancelling all inflight requests and
+// StopWithWait gracefully shuts down the producer, allowing all inflight
+// requests to be sent and for all backend connections to be closed. Once those
+// actions are complete, the wait group is signaled.
+//
+// It is safe to call the method, or Stop, multiple times and from multiple
+// goroutines; they will all block until the producer has been completely
+// shutdown. Note, though, that if Stop is called first, a subsequent call to
+// StopWithWait will not allow for inflight requests to be sent.
+func (p *Producer) StopWithWait(wg *sync.WaitGroup) {
+	go func() {
+		defer wg.Done()
+
+		p.once.Do(func() {
+			close(p.reqs)
+		})
+		p.join.Wait()
+	}()
+}
```

Comment:
Passing a wait group as argument is a bit of an unusual pattern in a Go library, starting goroutines and orchestrating to wait is often better handled in the main program.

How about revisiting to something like this:
```suggestion
func (p *Producer) Shutdown() {
	p.once.Do(func() {
		close(p.reqs)
	})
	p.join.Wait()
}
```
Then wherever a program calls `Shutdown`, it can add the async-wait behavior:
```go
wg := sync.WaitGroup{}

wg.Add(1)
go func() {
	defer wg.Done()
	p.Shutdown()
}()

// do other things

wg.Wait()
```

Reply:
Great, thanks for this advice! I'm still to Go so I don't know the "usual patterns" very well. I'll adjust the design.